README.md
# MAuth Client

This gem consists of MAuth::Client, a class to manage the information needed to both sign and authenticate requests
and responses, and middlewares for Rack and Faraday which leverage the client's capabilities.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'mauth-client'
```

And then execute:
```
$ bundle
```

Or install it yourself as:
```
$ gem install mauth-client
```


## Configuration

MAuth is typically configured by a yaml file, [mauth.yml](doc/mauth.yml.md) - see its page for more documentation.
This is simply loaded and passed to either middleware or directly to a MAuth::Client instance.
See the documentation for [MAuth::Client#initialize](lib/mauth/client.rb) for more details of what it accepts. Usually you will want:

```ruby
mauth_config = MAuth::Client.default_config
```

The `.default_config` method takes a number of options to tweak its expectations regarding defaults. See the
documentation for [MAuth::Client.default_config](lib/mauth/client.rb) for details.

The `private_key` and `app_uuid` (which go in mauth.yml) enable local authentication (see section [Local Authentication](#local-authentication) below).
They’ll only work if the `app_uuid` has been stored in MAuth with a public key corresponding to the `private_key` in mauth.yml.

If you do not have an `app_uuid` and keypair registered with the mauth service, you can use mauth's remote request authentication by omitting those fields.
MAuth-Client will make a call to MAuth for every request in order to authenticate remotely.
Remote authentication therefore requires more time than local authentication.
You will not be able to sign your responses without an `app_uuid` and a private key, so `MAuth::Rack::ResponseSigner` cannot be used.

The `mauth_baseurl` and `mauth_api_version` are required in mauth.yml.
These tell the MAuth Client where and how to communicate with the MAuth service.


## Rack Middleware Usage

MAuth Client provides a middleware for request authentication and response verification in mauth/rack.

```ruby
require 'mauth/rack'
```

If you are using other rack middlewares, the MAuth middleware MUST come FIRST in the stack of middlewares.
This means it is closest to the HTTP layer, furthest from the application.
If any other middlewares which modify the incoming request or outgoing response lie between the HTTP layer and the MAuth middleware, incoming requests will probably fail to authenticate and outgoing response signatures will be invalid (and fail when the requester tries to authenticate them).

Using these middlewares in rails consists of calls to `config.middleware.use` in the appropriate place (see [the Rails Guides](http://guides.rubyonrails.org/rails_on_rack.html) for more info).

Using the `MAuth::Rack::ResponseSigner` middleware is optional, but highly recommended.
If used, this should come before the `MAuth::Rack::RequestAuthenticator` middleware.
The ResponseSigner can be used ONLY if you have an `app_uuid` and `private_key` specified in your mauth configuration.

```ruby
config.middleware.use MAuth::Rack::ResponseSigner, mauth_config
```

Then request authentication:

```ruby
config.middleware.use MAuth::Rack::RequestAuthenticator, mauth_config
```

However, assuming you have a route `/app_status`, you probably want to skip request authentication for that.
There is a middleware (`RequestAuthenticatorNoAppStatus`) to make that easier:

```ruby
config.middleware.use MAuth::Rack::RequestAuthenticatorNoAppStatus, mauth_config
```

You may want to configure other conditions in which to bypass MAuth authentication.
The middleware takes an option on the `:should_authenticate_check` key, which is a ruby proc that is passed to the request's rack env and must result in a boolean.
If the result is true(ish), the middleware will authenticate the incoming request; if false, it will not.
The `:should_authenticate_check` parameter is OPTIONAL.
If omitted, all incoming requests will be authenticated.

Here are a few example `:should_authenticate_check` procs:

```ruby
mauth_config[:should_authenticate_check] = proc do |env|
  env['REQUEST_METHOD'] == 'GET'
end
config.middleware.use MAuth::Rack::RequestAuthenticator, mauth_config
```

Above, env is a hash of request parameters; this hash is generated by Rack.
The above proc will force the middleware to authenticate only GET requests.


Another example:

```ruby
mauth_config[:should_authenticate_check] = proc do |env|
  env['PATH_INFO'] == '/studies.json'
end
config.middleware.use MAuth::Rack::RequestAuthenticator, mauth_config
```

The above proc will force the rack middleware to authenticate only requests to the "/studies.json" path.
To authenticate a group of related URIs, considered matching `env['PATH_INFO']` with one or more regular expressions.

The configuration passed to the middlewares in the above examples (`mauth_config`) is used create a new instance of `MAuth::Client`.
If you are managing an MAuth::Client of your own for some reason, you can pass that in on the key `:mauth_client => your_client`, and omit any other MAuth::Client configuration.
`:should_authenticate_check` is handled by the middleware and should still be specified alongside `:mauth_client`, if you are using it.

When the request authentication middleware determines that a request is inauthentic, it will not call the application and will respond with a 401 status code along with an error, expressed in JSON
(Content-Type: application/json) with the following value:
```
{ "errors": { "mauth": ["Unauthorized"] } }
```
Successfully authenticated requests will be passed to the application, as will requests for which the `:should_authenticate_check` condition is false.

If the middleware is unable to authenticate the request because MAuth is unavailable and so cannot serve public keys, it responds with a 500 status code and an error expressed in JSON with the value:
```
{ "errors": { "mauth": ["Could not determine request authenticity"] } }
```

## Examples

Putting all this together, here are typical examples (in rails you would put that code in an initializer):

```ruby
mauth_config = MAuth::Client.default_config
require 'mauth/rack'
config.middleware.use MAuth::Rack::ResponseSigner, mauth_config
config.middleware.use MAuth::Rack:: RequestAuthenticatorNoAppStatus, mauth_config
```

With `:should_authenticate_check`:

```ruby
mauth_config = MAuth::Client.default_config
require 'mauth/rack'
config.middleware.use MAuth::Rack::ResponseSigner, mauth_config
# authenticate all requests which pass the some_condition_of check and aren't /app_status with MAuth
mauth_config[:should_authenticate_check] = proc do |env|
  some_condition_of(env)
end
config.middleware.use MAuth::Rack:: RequestAuthenticatorNoAppStatus, mauth_config
```

## Fake middleware

For testing purposes, you may wish to use middleware which does not perform actual authentication.
MAuth provides this, as `MAuth::Rack::RequestAuthenticationFaker`.
Requests are still checked for the presence of an MAuth signature - this is necessary as many applications rely on the `app_uuid` identified in the signature, so it cannot be ignored entirely.
However, the validity of the public key is not checked in the MAuth service, and the authenticity of the request is not verified by its signature.

This example code may augment the above examples to disable authentication in test mode:

```ruby
require 'mauth/fake/rack'
authenticator = Rails.env != 'test' ? MAuth::Rack::RequestAuthenticator : MAuth::Rack::RequestAuthenticationFaker
config.middleware.use authenticator, mauth_config
```

## Faraday Middleware Usage

If you are making outgoing HTTP requests using Faraday, adding MAuth Faraday middleware is much the same as adding rack middleware.
Building your connection will look like:

```ruby
Faraday.new(some_args) do |builder|
  builder.use MAuth::Faraday::RequestSigner, mauth_config
  builder.use MAuth::Faraday::ResponseAuthenticator, mauth_config
  builder.adapter Faraday.default_adapter
end
```

The Faraday middleware MUST come LAST in the stack of middleware.
As with the rack middleware, this means it will be right next to the HTTP adapter.

Only use the `MAuth::Faraday::ResponseAuthenticator` middleware if you are expecting the service you are communicating with to sign its responses (all services which are aware of MAuth _should_ be doing this).

`mauth_config` is the same as in Rack middleware, and as with the Rack middleware is used to initialize a `MAuth::Client` instance.
Also as with the Rack middleware, you can pass in a `MAuth::Client` instance you are using yourself on the `:mauth_client` key, and omit any other configuration.

Behavior is likewise similar to rack: if a `private_key` and `app_uuid` are specified, then ResponseAuthenticator will authenticate locally (see [Local Authentication](#local-authentication) below); if not, then it will go to the
mauth service to authenticate.
`MAuth::Faraday::RequestSigner` cannot be used without a `private_key` and `app_uuid`.

If a response which does not appear to be authentic is received by the `MAuth::Faraday::ResponseAuthenticator` middleware, a `MAuth::InauthenticError` will be raised.

If the MAuth service cannot be reached, and therefore the authenticity of a response cannot be verified by ResponseAuthenticator, then a `MAuth::UnableToAuthenticateError` will be raised.

## Other Request and Response signing

If you are not using Faraday, you will need to sign your own requests.

Instantiate a `MAuth::Client` with the same configuration as the middlewares, as documented on [MAuth::Client#initialize](lib/mauth/client.rb).
We'll call this `mauth_client`.

`mauth_client` has a method `#signed_headers` which takes either a `MAuth::Request` or `MAuth::Response` object, and generates HTTP headers which can be added to the request or response to indicate authenticity.
Create a `MAuth::Request` object from the information in your HTTP request, whatever its form:

```ruby
require 'mauth/request_and_response'
request = MAuth::Request.new(verb: my_verb, request_url: my_request_url, body: my_body)
```
`mauth_client.signed_headers(request)` will then return mauth headers which you can apply to your request.

## Local Authentication

When doing local authentication, the mauth client will periodically fetch and cache public keys from MAuth.
Each public key will be cached locally for 60 seconds.
Applications which connect frequently to the app will benefit most from this caching strategy.
When fetching public keys from MAuth, the following rules apply:

1. If MAuth returns the public key for a given `app_uuid`, MAuth-Client will refresh its local cache with this new public key.
2. If MAuth cannot find the public key for a given `app_uuid` (i.e. returns a 404 status code), MAuth-Client will remove the corresponding public key from its local cache and authentication of any message from the application with this public key will fail as a consequence.
3. If the request to MAuth times out or MAuth returns a 500 status code, the requested public key will not be removed from local MAuth-Client cache (if it exists there in the first place).
   The cached version will continue to be used for local authentication until MAuth::Client is able to again communicate with MAuth.

## Warning

During development classes are typically not cached in Rails applications.
If this is the case, be aware that the MAuth-Client middleware object will be instantiated anew for each request;
this will cause applications performing local authentication to fetch public keys before each request is authenticated.


History.md
# MAuth-Client History

## 2.4.0

- colorized output from the mauth-client CLI 
- add --content-type option to CLI
- improved method documentation 
- fix default null logger on windows where /dev/null is not available 
- improve error logging

## 2.3.0

- when authentication headers are missing, the previous message ("No x-mws-time present") is replaced by the somewhat 
  more informative "Authentication Failed. No mAuth signature present; X-MWS-Authentication header is blank."
- more informative help messages from mauth-client CLI
- CLI sets a user-agent 
- handling timeout errors is fixed (previously only handled connection errors)
- middleware MAuth::Rack::RequestAuthenticationFaker for testing 
- more and better specs 

## v2.2.0:

- fixes an issue where requests which have a body and are not PUT or POST were not being correctly signed in rack 
  middleware 
- improves the CLI, adding command-line options --[no-]authenticate to decide whether to authenticate responses, and 
  --[no-]verbose to decide whether to dump the entire request and response, or just the response body. and --help to 
  remind you. 
- fixes mauth-client CLI being registered as an executable in the gemspec - now it should be possible to just 
  `bundle exec mauth-client` if you have the gem bundle installed (or just `mauth-client` if you have it installed as 
  a regular gem, but that's less straightforward) 
- new middleware MAuth::Rack::RequestAuthenticatorNoAppStatus - same as MAuth::Rack::RequestAuthenticator, but does 
  not authenticate /app_status. this will be the most commonly used case, so made it its own middleware. 
- middleware responds to HEAD requests correctly in error conditions, not including a response body 
- drops backports dependency (Ben has found some issues with this gem, and it was easier to drop the depedency 
  entirely than figure out whether these issues affected mauth-client and if it could be fixed) 
- fix issue with remote authentication against the currently-deployed mauth service with a request signed by a 
  nonexistent app_uuid

## v2.1.1

- fix an issue in a case where the rack.input is not rewound before mauth-client attempts to read it

## v2.1.0

- MAuth::Client handles the :private_key_file, so you can remove from your application the bit that does that - this 
  bit can be deleted:

<pre>
if mauth_conf['private_key_file']
  mauth_conf['private_key'] = File.read(mauth_conf['private_key_file'])
end
</pre>

- autoloads are in place so that once you require 'mauth/client', you should not need to require mauth/rack, 
  mauth/faraday, or mauth/request_and_response.

## v2.0.0

- Rewrite combining the mauth_signer and rack-mauth gems 


lib/mauth/client.rb
require 'uri'
require 'openssl'
require 'base64'
require 'json'

require 'mauth/core_ext'
require 'mauth/autoload'

module MAuth
  class Client
    class << self
      # the root of the MAuth-Client library 
      def root
        File.expand_path('../..', File.dirname(__FILE__))
      end

      # the MAuth::Client library's current version 
      def version
        version_file = File.join(root, 'VERSION')
        File.exists?(version_file) ? File.read(version_file).chomp : "?"
      end
    end
  end
end

module MAuth
  # mAuth client was unable to verify the authenticity of a signed object (this does NOT mean the 
  # object is inauthentic). typically due to a failure communicating with the mAuth service, in 
  # which case the error may include the attribute mauth_service_response - a response from 
  # the mauth service (if it was contactable at all), which may contain more information about 
  # the error. 
  class UnableToAuthenticateError < StandardError
    # the response from the MAuth service encountered when attempting to retrieve authentication 
    attr_accessor :mauth_service_response
  end

  # used to indicate that an object was expected to be validly signed but its signature does not 
  # match its contents, and so is inauthentic. 
  class InauthenticError < StandardError
  end

  # required information for signing was missing 
  class UnableToSignError < StandardError
  end

  # does operations which require a private key and corresponding app uuid. this is primarily:
  # - signing outgoing requests and responses 
  # - authenticating incoming requests and responses, which may require retrieving the appropriate 
  #   public key from mAuth (which requires a request to mAuth which is signed using the private 
  #   key) 
  #
  # this nominally operates on request and response objects, but really the only requirements are 
  # that the object responds to the methods of MAuth::Signable and/or MAuth::Signed (as 
  # appropriate) 
  class Client
    MWS_TOKEN = 'MWS'

    # new client with the given App UUID and public key. config may include the following (all 
    # config keys may be strings or symbols):
    # - private_key - required for signing and for authenticating responses. may be omitted if 
    #   only remote authentication of requests is being performed (with 
    #   MAuth::Rack::RequestAuthenticator). may be given as a string or a OpenSSL::PKey::RSA 
    #   instance. 
    # - app_uuid - required in the same circumstances where a private_key is required 
    # - mauth_baseurl - required. needed for local authentication to retrieve public keys; needed 
    #   for remote authentication for hopefully obvious reasons. 
    # - mauth_api_version - required. only 'v1' exists / is supported as of this writing. 
    # - logger - a Logger to which any useful information will be written. if this is omitted and 
    #   Rails.logger exists, that will be used. 
    # - authenticator - this pretty much never needs to be specified. LocalAuthenticator or 
    #   RemoteRequestAuthenticator will be used as appropriate. 
    def initialize(config={})
      # stringify symbol keys
      given_config = config.stringify_symbol_keys
      # build a configuration which discards any irrelevant parts of the given config (small memory usage matters here) 
      @config = {}
      if given_config['private_key_file'] && !given_config['private_key']
        given_config['private_key'] = File.read(given_config['private_key_file'])
      end
      @config['private_key'] = case given_config['private_key']
      when nil
        nil
      when String
        OpenSSL::PKey::RSA.new(given_config['private_key'])
      when OpenSSL::PKey::RSA
        given_config['private_key']
      else
        raise ArgumentError, "unrecognized value given for 'private_key' - this may be a String, a OpenSSL::PKey::RSA, or omitted; instead got: #{given_config['private_key'].inspect}"
      end
      @config['app_uuid'] = given_config['app_uuid']
      @config['mauth_baseurl'] = given_config['mauth_baseurl']
      @config['mauth_api_version'] = given_config['mauth_api_version']
      @config['logger'] = given_config['logger'] || begin
        if Object.const_defined?('Rails')
          Rails.logger
        else
          require 'logger'
          is_win = RUBY_PLATFORM =~ /mswin|windows|mingw32|cygwin/i
          null_device = is_win ? 'NUL' : '/dev/null'
          ::Logger.new(File.open(null_device, File::WRONLY))
        end
      end
      @config['faraday_options'] = {:timeout => 1, :open_timeout => 1}.merge(given_config['faraday_options'] || {})

      # if 'authenticator' was given, don't override that - including if it was given as nil / false 
      if given_config.key?('authenticator')
        @config['authenticator'] = given_config['authenticator']
      else
        if client_app_uuid && private_key
          # MAuth::Client can authenticate locally if it's provided a client_app_uuid and private_key
          @config['authenticator'] = LocalAuthenticator
        else
          # otherwise, it will authenticate remotely (requests only) 
          @config['authenticator'] = RemoteRequestAuthenticator
        end
      end
      extend @config['authenticator'] if @config['authenticator']
    end

    def logger
      @config['logger']
    end
    def client_app_uuid
      @config['app_uuid']
    end
    def mauth_baseurl
      @config['mauth_baseurl'] || raise("no configured mauth_baseurl!")
    end
    def mauth_api_version
      @config['mauth_api_version'] || raise("no configured mauth_api_version!")
    end
    def private_key
      @config['private_key']
    end
    def faraday_options
      @config['faraday_options']
    end
    def assert_private_key(err)
      unless private_key
        raise err
      end
    end

    private
    def mauth_service_response_error(response)
      message = "mAuth service responded with #{response.status}: #{response.body}"
      logger.error(message)
      error = UnableToAuthenticateError.new(message)
      error.mauth_service_response = response || response_error
      raise error
    end

    # methods to sign requests and responses. part of MAuth::Client 
    module Signer
      # takes an outgoing request or response object, and returns an object of the same class 
      # whose headers are updated to include mauth's signature headers 
      def signed(object, attributes={})
        object.merge_headers(signed_headers(object, attributes))
      end

      # takes a signable object (outgoing request or response). returns a hash of headers to be 
      # applied tothe object which comprise its signature. 
      def signed_headers(object, attributes={})
        attributes = {:time => Time.now.to_i.to_s, :app_uuid => client_app_uuid}.merge(attributes)
        signature = self.signature(object, attributes)
        {'X-MWS-Authentication' => "#{MWS_TOKEN} #{client_app_uuid}:#{signature}", 'X-MWS-Time' => attributes[:time]}
      end

      # takes a signable object (outgoing request or response). returns a mauth signature string 
      # for that object. 
      def signature(object, attributes={})
        assert_private_key(UnableToSignError.new("mAuth client cannot sign without a private key!"))
        attributes = {:time => Time.now.to_i.to_s, :app_uuid => client_app_uuid}.merge(attributes)
        signature = Base64.encode64(private_key.private_encrypt(object.string_to_sign(attributes))).gsub("\n","")
      end
    end
    include Signer

    # methods common to RemoteRequestAuthenticator and LocalAuthenticator 
    module Authenticator
      ALLOWED_DRIFT_SECONDS = 300

      # takes an incoming request or response object, and returns whether 
      # the object is authentic according to its signature. 
      def authentic?(object)
        begin
          authenticate!(object)
          true
        rescue InauthenticError
          false
        end
      end

      # raises InauthenticError unless the given object is authentic 
      def authenticate!(object)
        authentication_present!(object)
        time_valid!(object)
        token_valid!(object)
        signature_valid!(object)
      rescue InauthenticError
        logger.error "mAuth signature authentication failed for #{object.class}. encountered error:"
        $!.message.split("\n").each{|l| logger.error "\t#{l}" }
        raise
      rescue UnableToAuthenticateError
        logger.error "Unable to authenticate with MAuth. encountered error:"
        $!.message.split("\n").each{|l| logger.error "\t#{l}" }
        raise
      end

      private
      def authentication_present!(object)
        if object.x_mws_authentication.nil? || object.x_mws_authentication !~ /\S/
          raise InauthenticError, "Authentication Failed. No mAuth signature present; X-MWS-Authentication header is blank."
        end
      end

      def time_valid!(object, now=Time.now)
        if object.x_mws_time.nil?
          raise InauthenticError, "Time verification failed for #{object.class}. No x-mws-time present."
        elsif !(-ALLOWED_DRIFT_SECONDS..ALLOWED_DRIFT_SECONDS).include?(now.to_i - object.x_mws_time.to_i)
          raise InauthenticError, "Time verification failed for #{object.class}. #{object.x_mws_time} not within #{ALLOWED_DRIFT_SECONDS} of #{now}"
        end
      end
      def token_valid!(object)
        unless object.signature_token==MWS_TOKEN
          raise InauthenticError, "Token verification failed for #{object.class}. Expected #{MWS_TOKEN.inspect}; token was #{object.signature_token}"
        end
      end
    end
    include Authenticator

    # methods to verify the authenticity of signed requests and responses locally, retrieving 
    # public keys from the mAuth service as needed 
    module LocalAuthenticator
      private
      def signature_valid!(object)
        expected = object.string_to_sign(:time => object.x_mws_time, :app_uuid => object.signature_app_uuid)
        pubkey = OpenSSL::PKey::RSA.new(retrieve_public_key(object.signature_app_uuid))
        begin
          actual = pubkey.public_decrypt(Base64.decode64(object.signature))
        rescue OpenSSL::PKey::PKeyError
          raise InauthenticError, "Public key decryption of signature failed!\n#{$!.class}: #{$!.message}"
        end
        # TODO: time-invariant comparison instead of #== ? 
        unless expected == actual
          raise InauthenticError, "Signature verification failed for #{object.class}"
        end
      end
      def retrieve_public_key(app_uuid)
        retrieve_security_token(app_uuid)['security_token']['public_key_str']
      end
      def retrieve_security_token(app_uuid)
        security_token_cacher.get(app_uuid)
      end
      def security_token_cacher
        @security_token_cacher ||= SecurityTokenCacher.new(self)
      end
      class SecurityTokenCacher
        class ExpirableSecurityToken < Struct.new(:security_token, :create_time)
          CACHE_LIFE=60
          def expired?
            create_time + CACHE_LIFE < Time.now
          end
        end
        def initialize(mauth_client)
          @mauth_client = mauth_client
          @mauth_client.assert_private_key(UnableToAuthenticateError.new("Cannot fetch public keys from mAuth service without a private key!")) # TODO should this be UnableToSignError? 
          @cache = {}
          require 'thread'
          @cache_write_lock = Mutex.new
        end
        def get(app_uuid)
          if !@cache[app_uuid] || @cache[app_uuid].expired?
            # url-encode the app_uuid to prevent trickery like escaping upward with ../../ in a malicious 
            # app_uuid - probably not exploitable, but this is the right way to do it anyway. 
            # use UNRESERVED instead of UNSAFE (the default) as UNSAFE doesn't include / 
            url_encoded_app_uuid = URI.escape(app_uuid, Regexp.new("[^#{URI::PATTERN::UNRESERVED}]"))
            begin
              response = signed_mauth_connection.get("/mauth/#{@mauth_client.mauth_api_version}/security_tokens/#{url_encoded_app_uuid}.json")
            rescue ::Faraday::Error::ConnectionFailed, ::Faraday::Error::TimeoutError
              raise UnableToAuthenticateError, "mAuth service did not respond; received #{$!.class}: #{$!.message}"
            end
            if response.status==200
              begin
                security_token = JSON.parse(response.body)
              rescue JSON::ParserError
                raise UnableToAuthenticateError, "mAuth service responded with unparseable json: #{response.body}\n#{$!.class}: #{$!.message}"
              end
              @cache_write_lock.synchronize do
                @cache[app_uuid] = ExpirableSecurityToken.new(security_token, Time.now)
              end
            elsif response.status==404
              # signing with a key mAuth doesn't know about is considered inauthentic 
              raise InauthenticError, "mAuth service responded with 404 looking up public key for #{app_uuid}"
            else
              @mauth_client.send(:mauth_service_response_error, response)
            end
          end
          @cache[app_uuid].security_token
        end

        private
        def signed_mauth_connection
          require 'faraday'
          require 'mauth/faraday'
          @signed_mauth_connection ||= ::Faraday.new(@mauth_client.mauth_baseurl, @mauth_client.faraday_options) do |builder|
            builder.use MAuth::Faraday::MAuthClientUserAgent
            builder.use MAuth::Faraday::RequestSigner, 'mauth_client' => @mauth_client
            builder.adapter ::Faraday.default_adapter
          end
        end
      end
    end

    # methods for remotely authenticating a request by sending it to the mauth service 
    module RemoteRequestAuthenticator
      private

      # takes an incoming request object (no support for responses currently), and errors if the 
      # object is not authentic according to its signature 
      def signature_valid!(object)
        raise ArgumentError, "Remote Authenticator can only authenticate requests; received #{object.inspect}" unless object.is_a?(MAuth::Request)
        authentication_ticket = {
          'verb' => object.attributes_for_signing[:verb],
          'app_uuid' => object.signature_app_uuid,
          'client_signature' => object.signature,
          'request_url' => object.attributes_for_signing[:request_url],
          'request_time' => object.x_mws_time,
          'b64encoded_body' => Base64.encode64(object.attributes_for_signing[:body] || '')
        }
        begin
          response = mauth_connection.post("/mauth/#{mauth_api_version}/authentication_tickets.json", {"authentication_ticket" => authentication_ticket})
        rescue ::Faraday::Error::ConnectionFailed, ::Faraday::Error::TimeoutError
          raise UnableToAuthenticateError, "mAuth service did not respond; received #{$!.class}: #{$!.message}"
        end
        if (200..299).include?(response.status)
          nil
        elsif response.status==412 || response.status==404
          # the mAuth service responds with 412 when the given request is not authentically signed. 
          # older versions of the mAuth service respond with 404 when the given app_uuid
          # does not exist, which is also considered to not be authentically signed. newer 
          # versions of the service respond 412 in all cases, so the 404 check may be removed 
          # when the old version of the mAuth service is out of service. 
          raise InauthenticError, "The mAuth service responded with #{response.status}: #{response.body}"
        else
          mauth_service_response_error(response)
        end
      end

      def mauth_connection
        require 'faraday'
        require 'faraday_middleware'
        @mauth_connection ||= ::Faraday.new(mauth_baseurl, faraday_options) do |builder|
          builder.use MAuth::Faraday::MAuthClientUserAgent
          builder.use FaradayMiddleware::EncodeJson
          builder.adapter ::Faraday.default_adapter
        end
      end
    end
  end
end
